crates/mvr-api/src/data/name_analytics.rs

```diff
@@ -0,0 +1,166 @@
+
+use super::reader::Reader;
+
+/// The `NameAnalyticsKey` expects a "Name" (MVR name),
+/// and an `ObjectId`, which can be ANY version of the package
+/// the name resolves to. This can be retrieved from the `package_by_name_loader`.
+///
+/// The `NaiveDate` is the date of the analytics query, as we are caching the query on a daily basis
+/// per instance.
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct NameAnalyticsKey(pub Name, pub ObjectId, pub NaiveDate);
+
+#[derive(Serialize, Deserialize, Clone, QueryableByName)]
+pub struct AnalyticsQueryResponse {
+    #[diesel(sql_type = diesel::sql_types::Text)]
+    pub package_id: String,
+
+    #[diesel(sql_type = diesel::sql_types::Date)]
+    pub date_from: NaiveDate,
+    #[diesel(sql_type = diesel::sql_types::Date)]
+    pub date_to: NaiveDate,
+    #[diesel(sql_type = diesel::sql_types::BigInt)]
+    pub direct: i64,
+    #[diesel(sql_type = diesel::sql_types::BigInt)]
+    pub propagated: i64,
+    #[diesel(sql_type = diesel::sql_types::BigInt)]
+    pub total: i64,
+}
+
+#[derive(Serialize, Deserialize, Clone, Default)]
+pub struct AnalyticsValue {
+    pub date_from: NaiveDate,
+    pub date_to: NaiveDate,
+    pub direct: i64,
+    pub propagated: i64,
+    pub total: i64,
+}
+
+#[derive(Serialize, Deserialize, Clone, Default)]
+pub struct AnalyticsAggregatedValues {
+    pub analytics: Vec<AnalyticsValue>,
+}
+
+#[async_trait::async_trait]
+impl Loader<NameAnalyticsKey> for Reader {
+    type Value = AnalyticsAggregatedValues;
+    type Error = ApiError;
+
+    async fn load(
+        &self,
+        keys: &[NameAnalyticsKey],
+    ) -> Result<HashMap<NameAnalyticsKey, Self::Value>, Self::Error> {
+        if keys.is_empty() {
+            return Ok(HashMap::new());
+        }
+
+        let mut connection = self.connect().await?;
+
+        let package_ids = keys.iter().map(|k| k.1.to_string()).collect::<Vec<_>>();
+
+        let query = diesel::sql_query(ANALYTICS_QUERY).bind::<Array<Text>, _>(package_ids);
+
+        let result: Vec<AnalyticsQueryResponse> = connection.results(query).await?;
+
+        let addr_name_mapping = keys.iter().map(|k| (k.1, k)).collect::<HashMap<_, _>>();
+
+        let mut aggregated_values: HashMap<_, _> = keys
+            .iter()
+            .map(|k| {
+                (
+                    k.clone(),
+                    AnalyticsAggregatedValues {
+                        analytics: Vec::new(),
+                    },
+                )
+            })
+            .collect();
+
+        for res in result.into_iter() {
+            // SAFETY: We should never have a malformed package id in the database.
+            let obj_id = ObjectId::from_str(&res.package_id).unwrap();
+
+            let Some(key) = addr_name_mapping.get(&obj_id) else {
+                continue;
+            };
+
+            aggregated_values.get_mut(key).and_then(|v| {
+                v.analytics.push(res.into());
+                Some(v)
+            });
```

Comment:
nit: if let ... is easier to read then else... continue ?

```suggestion
            if let Some(key) = addr_name_mapping.get(&obj_id) {
                if let Some(v) = aggregated_values.get_mut(key) {
                    v.analytics.push(res.into());
                };
            };
```